src/occupation.jl

```diff
@@ -64,50 +121,24 @@ function compute_fermi_level(basis::PlaneWaveBasis{T}, eigenvalues;
-                  "a temperature or switch to a calculation with collinear spin polarization.")
-        end
+    objective(εF) = deviation_n_electrons(basis, eigenvalues, εF; smearing, temperature)
+    @assert objective(min_ε) < 0 < objective(max_ε)
+    Roots.find_zero(objective, (min_ε, max_ε), Roots.Bisection(), atol=eps(T))
+end
 
-        # For zero temperature, two cases arise: either there are as many bands
-        # as electrons, in which case we set εF to the highest energy level
-        # reached, or there are unoccupied conduction bands and we take
-        # εF as the midpoint between valence and conduction bands.
-        if compute_n_elec(max_ε) ≈ n_electrons
-            εF = max_ε
-        else
-            # The sanity check above ensures that n_fill is well defined
-            n_fill = div(n_electrons, n_spin * filled_occ, RoundUp)
-            # highest occupied energy level
-            HOMO = maximum([εk[n_fill] for εk in eigenvalues])
-            HOMO = mpi_max(HOMO, basis.comm_kpts)
-            # lowest unoccupied energy level, be careful that not all k-points
-            # might have at least n_fill+1 energy levels so we have to take care
-            # of that by specifying init to minimum
-            LUMO = minimum(minimum.([εk[n_fill+1:end] for εk in eigenvalues]; init=T(Inf)))
-            LUMO = mpi_min(LUMO, basis.comm_kpts)
-            εF = (HOMO + LUMO) / 2
-        end
-    else
-        # For finite temperature, just use bisection; note that with MP smearing
-        # there might be multiple possible Fermi levels. This could be sped up
-        # with more advanced methods (e.g. false position), but more care has to
-        # be taken with convergence criteria and the like
-        @assert compute_n_elec(min_ε) < n_electrons < compute_n_elec(max_ε)
-        εF = Roots.find_zero(εF -> compute_n_elec(εF) - n_electrons, (min_ε, max_ε),
-                             Roots.Bisection(), atol=eps(T))
+function compute_fermi_level(basis::PlaneWaveBasis{T}, eigenvalues, ::GaussianNewton;
+                             temperature, smearing) where {T}
+    if iszero(temperature)
+        return compute_fermi_level(basis, eigenvalues, ZeroTemperature())
+    elseif smearing in (Smearing.Gaussian(), Smearing.FermiDirac())
+        # Monotonous smearing functions ... only one Fermi level anyway
+        return compute_fermi_level(basis, eigenvalues, Bisection(); temperature, smearing)
     end
 
-    if !isapprox(compute_n_elec(εF), n_electrons)
-        if iszero(temperature)
-            error("Unable to find non-fractional occupations that have the " *
-                  "correct number of electrons. You should add a temperature.")
-        else
-            error("This should not happen, debug me!")
-        end
-    end
-    εF
+    εF_guess = compute_fermi_level(basis, eigenvalues, Bisection();
+                                   temperature, smearing=Smearing.Gaussian())
```

Comment:
You could check for an early exit here if eF_guess happens to be right (not impossible).

Reply:
Indeed. Actually that's pretty likely even.